spec/system/csv_preview_spec.rb

```diff
@@ -0,0 +1,282 @@
+    before do
+      setup_asset_manager(parent_document_url, asset_manager_id, filename, malformed: true)
+      setup_content_item(asset_media_url_path, parent_document_base_path)
+      visit "/#{asset_media_url_path}/preview"
+    end
+
+    it "returns a 200 response" do
+      expect(page.status_code).to eq(200)
+    end
+
+    it "includes the error message" do
+      expect(page).to have_text("This CSV cannot be viewed online.")
+    end
+
+    context "when visiting the preview that redirects to other asset" do
+      before do
+        setup_asset_manager(parent_document_url, asset_manager_id, filename, malformed: true, asset_deleted: true)
+        setup_content_item(asset_media_url_path, parent_document_base_path)
+        visit "/#{asset_media_url_path}/preview"
+      end
+
+      it "returns a 410 response" do
+        expect(page.status_code).to eq(410)
+      end
+    end
+  end
+
+  context "when visiting the preview with nonconvertible characters in the CSV" do
+    before do
+      setup_asset_manager(parent_document_url, asset_manager_id, filename, nonconvertible: true)
+      setup_content_item(asset_media_url_path, parent_document_base_path)
+      visit "/#{asset_media_url_path}/preview"
+    end
+
+    it "returns a 200 response" do
+      expect(page.status_code).to eq(200)
+    end
+
+    it "includes the error message" do
+      expect(page).to have_text("This CSV cannot be viewed online.")
+    end
+  end
+
+  context "when the asset is draft and not served from the draft host" do
+    before do
+      asset_manager_response = { id: "https://asset-manager.dev.gov.uk/assets/foo", parent_document_url:, draft: true }
+      stub_asset_manager_has_an_asset(asset_manager_id, asset_manager_response, "/#{filename}.csv")
+      visit "/#{asset_media_url_path}/preview"
+    end
+
+    it "redirects to the draft assets host" do
+      expect(current_url).to eq("http://draft-assets.dev.gov.uk/#{asset_media_url_path}/preview")
+    end
+  end
+
+  context "when the asset does not exist" do
+    before do
+      asset_manager_id_2 = 9876
+      stub_asset_manager_does_not_have_an_asset(asset_manager_id_2)
+      visit "/media/#{asset_manager_id_2}/#{filename}-2.csv/preview"
+    end
+
+    it "returns a 404 response" do
+      expect(page.status_code).to eq(404)
+    end
+  end
+
+  context "when asset manager file name does not match content store file name" do
+    before do
+      asset_manager_filename = "2023-01-10_filename"
+      content_store_filename = "2023-01-11_filename"
+      path_from_filename = ->(base_name) { "media/#{asset_manager_id}/#{base_name}.csv" }
+      setup_asset_manager(parent_document_url, asset_manager_id, asset_manager_filename)
+      setup_content_item(path_from_filename[content_store_filename], parent_document_base_path)
+      @preview_path = "/#{path_from_filename[asset_manager_filename]}/preview"
+    end
+
+    it "redirect to parent" do
+      get @preview_path
```

Comment:
So! Basically, I think the solution here is to make this a request test. It's only really interested in the redirection, and because the scaffolding we'd need to set this up as a system test is quite awkward, it's easier to split it out as a request test (which is clearly what it was with the get method) and create an AssetManagerHelper so that we don't repeat the complex setup methods. I've done that now.

Reply:
Request spec is a good idea 👍 

But you've still got the instance variable storing the path. I looked at all the other request specs in this PR and there's only [one other place](https://github.com/alphagov/frontend/pull/4090/files#diff-cb0b18e6af1af6949d2ad7b83c52dcade8ed005d1cc86057e536fdd7e3f70297) we do `get @path`, everywhere else we're doing `get "/path-blah"`

So I guess the easiest fix would just be to do `get "/#{path_from_filename[asset_manager_filename]}/preview"`  but that's a bit of a confusing path to wrap your head around. Is this any better:

```
it "redirects to parent" do
  preview_path = "/#{path_from_filename[asset_manager_filename]}/preview"
  get preview_path

  expect(response).to redirect_to(parent_document_url)
end
```
Don't think it's a big deal though, so feel free to leave it as is if you prefer.

Reply:
Ah, the problem with that is that path_from_filename is a proc created in the before block, which isn't available in the it block. But if I make that a method (which is a bit less surprising anyway, then this works nicely.